File: internal/providers/openai/moderations.go
package openai

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"

	gw "github.com/everstacklabs/everstack/internal/lib/handlers/gateway"
)

// Moderate classifies content for policy violations.
func (p *Provider) Moderate(ctx context.Context, req gw.ModerationRequest) (gw.ModerationResponse, error) {
	if p.cfg.APIKey == "" {
		return gw.ModerationResponse{}, fmt.Errorf("openai api key not provided")
	}

	// Build request payload
	payload := make(map[string]interface{})

	// Use input string or inputs array
	if req.Input != "" {
		payload["input"] = req.Input
	} else if len(req.Inputs) > 0 {
		// Convert to OpenAI format
		inputs := make([]map[string]interface{}, len(req.Inputs))
		for i, inp := range req.Inputs {
			item := map[string]interface{}{
				"type": inp.Type,
			}
			if inp.Type == "text" {
				item["text"] = inp.Text
			} else if inp.Type == "image_url" && inp.ImageURL != nil {
				item["image_url"] = map[string]string{
					"url": inp.ImageURL.URL,
				}
			}
			inputs[i] = item
		}
		payload["input"] = inputs
	}

	if req.Model != "" {
		payload["model"] = req.Model
	}

	buf, _ := json.Marshal(payload)
	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, p.cfg.BaseURL+"/moderations", bytes.NewReader(buf))
	if err != nil {
		return gw.ModerationResponse{}, err
	}
	p.setHeaders(httpReq.Header)

	resp, err := p.client.Do(httpReq)
	if err != nil {
		return gw.ModerationResponse{}, err
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 300 {
		b, _ := io.ReadAll(resp.Body)
		return gw.ModerationResponse{}, errors.New("openai moderation error: " + string(b))
	}

	var parsed struct {
		ID      string `json:"id"`
		Model   string `json:"model"`
		Results []struct {
			Flagged    bool `json:"flagged"`
			Categories struct {
				Hate                  bool `json:"hate"`
				HateThreatening       bool `json:"hate/threatening"`
				Harassment            bool `json:"harassment"`
				HarassmentThreatening bool `json:"harassment/threatening"`
				Illicit               bool `json:"illicit"`
				IllicitViolent        bool `json:"illicit/violent"`
				SelfHarm              bool `json:"self-harm"`
				SelfHarmIntent        bool `json:"self-harm/intent"`
				SelfHarmInstructions  bool `json:"self-harm/instructions"`
				Sexual                bool `json:"sexual"`
				SexualMinors          bool `json:"sexual/minors"`
				Violence              bool `json:"violence"`
				ViolenceGraphic       bool `json:"violence/graphic"`
			} `json:"categories"`
			CategoryScores struct {
				Hate                  float64 `json:"hate"`
				HateThreatening       float64 `json:"hate/threatening"`
				Harassment            float64 `json:"harassment"`
				HarassmentThreatening float64 `json:"harassment/threatening"`
				Illicit               float64 `json:"illicit"`
				IllicitViolent        float64 `json:"illicit/violent"`
				SelfHarm              float64 `json:"self-harm"`
				SelfHarmIntent        float64 `json:"self-harm/intent"`
				SelfHarmInstructions  float64 `json:"self-harm/instructions"`
				Sexual                float64 `json:"sexual"`
				SexualMinors          float64 `json:"sexual/minors"`
				Violence              float64 `json:"violence"`
				ViolenceGraphic       float64 `json:"violence/graphic"`
			} `json:"category_scores"`
		} `json:"results"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&parsed); err != nil {
		return gw.ModerationResponse{}, err
	}

	result := gw.ModerationResponse{
		ID:    parsed.ID,
		Model: parsed.Model,
	}

	for _, r := range parsed.Results {
		result.Results = append(result.Results, gw.ModerationResult{
			Flagged: r.Flagged,
			Categories: gw.ModerationCategories{
				Hate:                  r.Categories.Hate,
				HateThreatening:       r.Categories.HateThreatening,
				Harassment:            r.Categories.Harassment,
				HarassmentThreatening: r.Categories.HarassmentThreatening,
				Illicit:               r.Categories.Illicit,
				IllicitViolent:        r.Categories.IllicitViolent,
				SelfHarm:              r.Categories.SelfHarm,
				SelfHarmIntent:        r.Categories.SelfHarmIntent,
				SelfHarmInstructions:  r.Categories.SelfHarmInstructions,
				Sexual:                r.Categories.Sexual,
				SexualMinors:          r.Categories.SexualMinors,
				Violence:              r.Categories.Violence,
				ViolenceGraphic:       r.Categories.ViolenceGraphic,
			},
			CategoryScores: gw.ModerationCategoryScores{
				Hate:                  r.CategoryScores.Hate,
				HateThreatening:       r.CategoryScores.HateThreatening,
				Harassment:            r.CategoryScores.Harassment,
				HarassmentThreatening: r.CategoryScores.HarassmentThreatening,
				Illicit:               r.CategoryScores.Illicit,
				IllicitViolent:        r.CategoryScores.IllicitViolent,
				SelfHarm:              r.CategoryScores.SelfHarm,
				SelfHarmIntent:        r.CategoryScores.SelfHarmIntent,
				SelfHarmInstructions:  r.CategoryScores.SelfHarmInstructions,
				Sexual:                r.CategoryScores.Sexual,
				SexualMinors:          r.CategoryScores.SexualMinors,
				Violence:              r.CategoryScores.Violence,
				ViolenceGraphic:       r.CategoryScores.ViolenceGraphic,
			},
		})
	}

	return result, nil
}

File: packages/sdks/everstack-node/src/resources/moderations.ts
/**
 * Moderations resource
 *
 * Provides content moderation / classification functionality.
 */

import type { Client } from "@connectrpc/connect";
import type { GatewayService } from "@everstack/proto/everstack/gateway/v1/gateway_service_pb.js";
import { create } from "@bufbuild/protobuf";
import {
  ModerationRequestSchema,
  type ModerationResponse as ProtoModerationResponse,
} from "@everstack/proto/everstack/gateway/v1/moderations_pb.js";

import { fromConnectError } from "../errors.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface ModerationParams {
  /** Text input to classify (shorthand for a single text input) */
  input?: string;
  /** Multiple typed inputs (text or image_url) */
  inputs?: Array<
    | { type: "text"; text: string }
    | { type: "image_url"; image_url: { url: string } }
  >;
  /** Model to use */
  model?: string;
}

export interface ModerationCategories {
  hate: boolean;
  "hate/threatening": boolean;
  harassment: boolean;
  "harassment/threatening": boolean;
  illicit: boolean;
  "illicit/violent": boolean;
  "self-harm": boolean;
  "self-harm/intent": boolean;
  "self-harm/instructions": boolean;
  sexual: boolean;
  "sexual/minors": boolean;
  violence: boolean;
  "violence/graphic": boolean;
}

export interface ModerationCategoryScores {
  hate: number;
  "hate/threatening": number;
  harassment: number;
  "harassment/threatening": number;
  illicit: number;
  "illicit/violent": number;
  "self-harm": number;
  "self-harm/intent": number;
  "self-harm/instructions": number;
  sexual: number;
  "sexual/minors": number;
  violence: number;
  "violence/graphic": number;
}

export interface ModerationResult {
  flagged: boolean;
  categories: ModerationCategories;
  category_scores: ModerationCategoryScores;
}

export interface ModerationResponse {
  id: string;
  model: string;
  results: ModerationResult[];
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function transformResponse(proto: ProtoModerationResponse): ModerationResponse {
  return {
    id: proto.id,
    model: proto.model,
    results: proto.results.map((r) => ({
      flagged: r.flagged,
      categories: {
        hate: r.categories?.hate ?? false,
        "hate/threatening": r.categories?.hateThreatening ?? false,
        harassment: r.categories?.harassment ?? false,
        "harassment/threatening": r.categories?.harassmentThreatening ?? false,
        illicit: r.categories?.illicit ?? false,
        "illicit/violent": r.categories?.illicitViolent ?? false,
        "self-harm": r.categories?.selfHarm ?? false,
        "self-harm/intent": r.categories?.selfHarmIntent ?? false,
        "self-harm/instructions": r.categories?.selfHarmInstructions ?? false,
        sexual: r.categories?.sexual ?? false,
        "sexual/minors": r.categories?.sexualMinors ?? false,
        violence: r.categories?.violence ?? false,
        "violence/graphic": r.categories?.violenceGraphic ?? false,
      },
      category_scores: {
        hate: r.categoryScores?.hate ?? 0,
        "hate/threatening": r.categoryScores?.hateThreatening ?? 0,
        harassment: r.categoryScores?.harassment ?? 0,
        "harassment/threatening": r.categoryScores?.harassmentThreatening ?? 0,
        illicit: r.categoryScores?.illicit ?? 0,
        "illicit/violent": r.categoryScores?.illicitViolent ?? 0,
        "self-harm": r.categoryScores?.selfHarm ?? 0,
        "self-harm/intent": r.categoryScores?.selfHarmIntent ?? 0,
        "self-harm/instructions": r.categoryScores?.selfHarmInstructions ?? 0,
        sexual: r.categoryScores?.sexual ?? 0,
        "sexual/minors": r.categoryScores?.sexualMinors ?? 0,
        violence: r.categoryScores?.violence ?? 0,
        "violence/graphic": r.categoryScores?.violenceGraphic ?? 0,
      },
    })),
  };
}

// ---------------------------------------------------------------------------
// Resource
// ---------------------------------------------------------------------------

/**
 * Moderations resource for content classification
 */
export class Moderations {
  /** @internal */
  constructor(private readonly _client: Client<typeof GatewayService>) {}

  /**
   * Classify content for policy violations
   *
   * @example
   * ```typescript
   * const result = await client.moderations.create({
   *   input: 'Some text to check',
   *   model: 'text-moderation-latest',
   * });
   * console.log(result.results[0].flagged);
   * ```
   */
  async create(params: ModerationParams): Promise<ModerationResponse> {
    try {
      const request = create(ModerationRequestSchema, {
        input: params.input ?? "",
        inputs: (params.inputs ?? []).map((i) => {
          if (i.type === "text") {
            return { type: "text", text: i.text, imageUrl: undefined };
          }
          return {
            type: "image_url",
            text: "",
            imageUrl: { url: i.image_url.url },
          };
        }),
        model: params.model ?? "",
      });

      const response = await this._client.moderation(request);
      return transformResponse(response);
    } catch (error) {
      throw fromConnectError(error);
    }
  }
}
